layers/+lang/c-c++/funcs.el

```diff
@@ -182,3 +185,162 @@ and the arguments for flyckeck-clang based on a project-specific text file."
+  (defun c-c++/member-hierarchy () (interactive) (spacemacs//c-c++-lsp-funcall-interactively-no-args nil "-member-hierarchy"))
+  (defun c-c++/freshen-index () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-freshen-index"))
+  (defun c-c++/preprocess-file () (interactive) (spacemacs//c-c++-lsp-funcall-interactively nil "-preprocess-file")))
+
+(defun spacemacs//c-c++-lsp-bind-keys-for-mode (mode)
+  "Bind LSP backend functions for the specificed mode."
+  (spacemacs/set-leader-keys-for-major-mode mode
+    ;; goto
+    ;; FIXME gf conflicts with `lsp-ui-flycheck-list' binding in lsp layer
+    ;; Better corrected in lsp-layer? rebinding here for now
+    "ge" #'lsp-ui-flycheck-list
+    "gf" 'find-file-at-point
+    "gF" 'ffap-other-window
+    "g&" #'c-c++/references-address
+    "gR" #'c-c++/references-read
+    "gW" #'c-c++/references-write
+    "gc" #'c-c++/callers
+    "gv" #'c-c++/vars
+    ;; help/hierarchy
+    "hb" #'c-c++/base ;;Replace this with lsp-goto-implementation in lsp-layer?
+    "hc" #'c-c++/call-hierarchy
+    "hC" #'c-c++/call-hierarchy-inv
+    "hi" #'c-c++/inheritance-hierarchy
+    "hI" #'c-c++/inheritance-hierarchy-inv
+    "hm" #'c-c++/member-hierarchy
+    ;; lsp/backend
+    "lf" #'c-c++/freshen-index
+    "lp" #'c-c++/preprocess-file))
+
+(defun spacemacs//c-c++-lsp-customise-lsp-ui-peek ()
+  "Wrap some backend-specific features using the `lsp-ui-peek' functions provided by the `lsp' layer (`lsp-ui' package)"
+  (spacemacs//c-c++-lsp-call-function "spacemacs//c-c++-lsp-" "-customise-lsp-ui-peek")
+
+  (defun c-c++/vars () (interactive) (lsp-ui-peek-find-custom 'vars (spacemacs//c-c++-lsp-string "$" "/vars")))
+
+  (defun c-c++/references-address ()
+    (interactive)
+    (lsp-ui-peek-find-custom
+      'address "textDocument/references"
+      (plist-put (lsp--text-document-position-params) :context
+        '(:role 128))))
+
+  (defun c-c++/references-read ()
+    (interactive)
+    (lsp-ui-peek-find-custom
+      'read "textDocument/references"
+      (plist-put (lsp--text-document-position-params) :context
+        '(:role 8))))
+
+  (defun c-c++/references-write ()
+    (interactive)
+    (lsp-ui-peek-find-custom
+      'write "textDocument/references"
+      (plist-put (lsp--text-document-position-params) :context
+        '(:role 16)))))
+
+(defun spacemacs//c-c++-lsp-cquery-customise-lsp-ui-peek ()
+  (defun c-c++/callers () (interactive) (lsp-ui-peek-find-custom 'callers "$cquery/callers"))
+  (defun c-c++/base () (interactive) (lsp-ui-peek-find-custom 'base "$cquery/base")))
+
+(defun spacemacs//c-c++-lsp-ccls-customise-lsp-ui-peek ()
+  (defun c-c++/callers () (interactive) (lsp-ui-peek-find-custom 'callers "$ccls/call"))
+  (defun c-c++/base ()
+    (interactive)
+    (lsp-ui-peek-find-custom 'base "$ccls/inheritance"
+      (append (lsp--text-document-position-params) '(:flat t :level 3)))))
```

Comment:
`c-c++/base` for ccls can be:

```elisp
(lsp-ui-peek-find-custom 'base "$ccls/inheritance" '(:levels 3))
```

`:levels 3` is for bases up to 3 levels above. It can be removed (`levels` defaults to 1) if direct bases are needed.

Reply:
Thanks fangrui -- I cleverly hadn't thought to recompile ccls in a while :)

Reply:
Since `$ccls/memberHierarchy` has been renamed to ``$ccls/member`, it makes me feel more comfortable to add other non hierarchy features, e.g. listing member functions of a class or listing nested classes/namespace members

```
(lsp-find-custom "$ccls/member" `(:kind 2)) => nested classes / types in a namespace
(lsp-find-custom "$ccls/member" `(:kind 3)) => member functions / functions in a namespace
(lsp-find-custom "$ccls/member" `(:kind 0)) => other than 2/3 => member variables / variables in a namespace
```

Diagnostics have also been improved much and memory footprint has been improved a lot.

Reply:
Thanks Fangrui. I've added bindings for them under `SPC m <g/p> m <c/f/v>` when ccls is selected as backend. Won't amend/push until #11351 gets merged/rejected, as I'm using those lsp-layer changes in my local c-c++ layer.